### PR TITLE
Java lab: Save to and Load from S3

### PR DESCRIPTION
**Pull request by molly-moen**
Enable loading and saving of a single file to/from S3 in Javalab. This entails a few things:
- on page load, try to find any existing files from the files api to load. Only load the first file for now and populate the editor with that file. If there are no files, give the editor a new file name (MyClass.java) and empty file contents.
- On rename, rename the file on S3
- On save, save the file to S3 (either add a new file or update an existing file)
- Editor code is only saved to the files system. In our sources system just save the files version id, like we do in Web Lab. The sources will get updated whenever we update a file, which will update the last saved time in the header.

Save is currently a manual process (need to click the button). Autosave will be added as a follow-up.

Minor other changes included here: 
- Changed "filename" to be all one word because it was a little inconsistent before.
- Updated the styling on codemirror to fill the entire containing div. If the code required horizontal scrolling, the scrollbar would previously appear right below the last line of code. Now it appears at the "bottom" of the editor.

New save button is labeled "commit code":
Screen Shot 2021-03-08 at 1 36 30 PM

I also improved the styling a bit on the rename save button: 
Screen Shot 2021-03-08 at 1 36 39 PM

## Links

- spec: [CSE IDE Design](https://docs.google.com/document/d/1m7s_DF2JM4ZRbbihj3HsuAA3n-wePegLvcwFQjEx2sA/edit)
- jira ticket: [CSA-122](https://codedotorg.atlassian.net/browse/CSA-122)

## Testing story
Tested locally. This work is still only available to levelbuilders and admins. Unit tests will come as a follow up along with auto-save

## Follow-up work
- Add autosave
- Add unit tests
- Add multi-file

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/javalab/Javalab.js
Comment (maddiedierker) [inline apps/src/javalab/Javalab.js]: string should say "saved," not "renamed"

File: apps/src/javalab/Javalab.js
Comment (maddiedierker) [inline apps/src/javalab/Javalab.js]: nit: make this error more specific (maybe something like, `FilesApi failed to get files. Status: ${xhr.status}`)

File: apps/src/javalab/Javalab.js
Comment (maddiedierker) [inline apps/src/javalab/Javalab.js]: right now, `<JavalabView/>` will still render as normal if we fail here. this may be future work, but what should happen if this fails? one idea would be to pass an error through to `<JavalabView/>` and render an error page instead of the normal view to let the user know their file couldn't be loaded.

File: apps/src/javalab/JavalabEditor.jsx
Comment (maddiedierker) [inline apps/src/javalab/JavalabEditor.jsx]: 2 nits: 
1. `activateRenameFile` might be a more accurate name?
2. some functions in this file are arrow functions and some aren't. i'm not sure what our team's current preference is (although i generally see us use arrow functions more often in React components), but we should be consistent

File: apps/src/javalab/JavalabEditor.jsx
Comment (maddiedierker) [inline apps/src/javalab/JavalabEditor.jsx:112]: possibly future work, but user-facing strings should live in an i18n file

File: apps/src/javalab/JavalabEditor.jsx
Comment (maddiedierker) [inline apps/src/javalab/JavalabEditor.jsx]: super nit: this can be simplified since it's a single command:

```
onChange={e => this.props.setFilename(e.target.value)}
```

File: apps/src/javalab/JavalabEditor.jsx
Comment (maddiedierker) [inline apps/src/javalab/JavalabEditor.jsx]: if `enableRenameFile` were an arrow function, this could simplify to `onClick={this.enableRenameFile}`

File: apps/src/javalab/Javalab.js
Comment (jmkulwik) [inline apps/src/javalab/Javalab.js]: I'm not super familiar with what these two lines do. Mind adding a comment?

File: apps/src/javalab/Javalab.js
Comment (jmkulwik) [inline apps/src/javalab/Javalab.js]: JS is not the best at telling a reader when/why an arrow function will be called. Mind adding a comment for these? Guessing it's something like `/* onSuccess */` and `/* onError */`?

File: apps/src/javalab/Javalab.js
Comment (jmkulwik) [inline apps/src/javalab/Javalab.js]: Should this call `saveIfSourcesChanged`?

File: apps/src/javalab/Javalab.js
Comment (jmkulwik) [inline apps/src/javalab/Javalab.js]: Rather than passing this down multiple levels as a callback, would it make more sense to make it a member function of Javalab and call it directly? Alternatively, consider renaming the `callback` parameter in `loadFiles`, `populateFiles`, and `requestFileEntryAndWrite` to something more descriptive.

File: apps/src/javalab/Javalab.js
Comment (jmkulwik) [inline apps/src/javalab/Javalab.js:138]: Would it make sense to handle file management in a separate module?

File: apps/src/javalab/JavalabEditor.jsx
Comment (jmkulwik) [reply apps/src/javalab/JavalabEditor.jsx]: Since this component has a constructor, I'd like to put a plug in for binding it in the constructor.

Reasoning: This is kinda obscure, but our [style guide](https://github.com/code-dot-org/code-dot-org/blob/staging/STYLEGUIDE.md#react) references the [airbnb style guide](https://github.com/airbnb/javascript/tree/master/react#methods) which has some compelling reasons for binding event handlers exactly once (i.e. it improves performance)
https://github.com/airbnb/javascript/tree/master/react#methods 

TL;DR it would look like this:
```
constructor(props) {
  super(props);
  this.enableRenameFile = this.enableRenameFile.bind(this);
  ...
}

enableRenameFile() {
  ...
}

...

onClick={this.enableRenameFile}
```

Comment (jmkulwik) [reply apps/src/javalab/JavalabEditor.jsx]: Similar for the other event handlers.

File: apps/style/javalab/style.scss
Comment (jmkulwik) [inline apps/style/javalab/style.scss:27]: In my shameless crusade to get rid of `!important` in our code base, I wanted to double check: Is `!important` truly important here?

File: apps/src/javalab/JavalabEditor.jsx
Comment (molly-moen) [reply apps/src/javalab/JavalabEditor.jsx:112]: 👍 right now  the strings aren't final so I'm not putting them in i18n, but will add them when we finalize them.

File: apps/style/javalab/style.scss
Comment (molly-moen) [reply apps/style/javalab/style.scss:27]: Yes, without `!important` there is a white border at the top of the page.

File: apps/src/javalab/Javalab.js
Comment (molly-moen) [reply apps/src/javalab/Javalab.js]: good call, I think I missed that.

File: apps/src/javalab/Javalab.js
Comment (molly-moen) [reply apps/src/javalab/Javalab.js]: it calls `save`, which doesn't check if the sources changed. The sources here is the files version id, which I think will always change. If we prefer `saveIfSourcesChanged` I can change it to that.

File: apps/src/javalab/Javalab.js
Comment (molly-moen) [reply apps/src/javalab/Javalab.js]: combining this feedback with your module idea below--refactoring the file management into a module and importing the functions into the required components (and using `success` and `failure` instead of `callback`). I think keeping all the file management in one place is a good idea because it will get more complicated with multi-file.

Comment (molly-moen) [reply apps/src/javalab/Javalab.js]: one downside of moving the load into the view is we are losing Studio App's default behavior of showing a "loading is taking a long time" message with a reload link if the page load takes a long time--we can discuss if we want to add that back or not, not super important right now.

File: apps/src/javalab/Javalab.js
Comment (molly-moen) [reply apps/src/javalab/Javalab.js]: did some refactoring of this--we didn't actually need to track initialFilesVersionId, so changed suppliedFilesVersionId to a const below and added a comment as to what it does.

File: apps/src/javalab/JavalabEditor.jsx
Comment (molly-moen) [reply apps/src/javalab/JavalabEditor.jsx]: updated!

File: apps/src/javalab/JavalabEditor.jsx
Comment (molly-moen) [reply apps/src/javalab/JavalabEditor.jsx]: renamed and bound the function in the constructor like Jessie suggested.

File: apps/src/javalab/Javalab.js
Comment (jmkulwik) [reply apps/src/javalab/Javalab.js]: 👍

File: apps/src/javalab/Javalab.js
Comment (jmkulwik) [reply apps/src/javalab/Javalab.js]: That makes sense. I forgot that this is the files api, not the sources api

File: apps/src/javalab/JavalabFileManagement.js
Comment (jmkulwik) [inline apps/src/javalab/JavalabFileManagement.js]: NIT: Still needed?

File: apps/src/javalab/JavalabFileManagement.js
Comment (jmkulwik) [inline apps/src/javalab/JavalabFileManagement.js]: NIT: Still needed?

File: apps/src/javalab/JavalabFileManagement.js
Comment (jmkulwik) [inline apps/src/javalab/JavalabFileManagement.js]: NIT: Still needed?

File: apps/src/javalab/Javalab.js
Comment (jmkulwik) [reply apps/src/javalab/Javalab.js]: That message would be nice to have, but I don't think it should block this PR. Why do we lose the message?

Comment (jmkulwik) [reply apps/src/javalab/Javalab.js]: Refactor looks great :)

**Review by jmkulwik — COMMENTED** [review]
Looks great! I left a couple nits, but they don't need to go through another review cycle.

File: apps/src/javalab/JavalabFileManagement.js
Comment (maddiedierker) [inline apps/src/javalab/JavalabFileManagement.js:109]: super nit: i think the more modern way to do this same thing is like this:

```javascript
export default {
  renameProjectFile,
  ...
}
```

File: apps/src/javalab/Javalab.js
Comment (molly-moen) [reply apps/src/javalab/Javalab.js]: the message pops up when the `init` function takes a certain amount of time--since we are loading after init completes now we lose the message.

File: apps/src/javalab/JavalabView.jsx
Comment (maddiedierker) [inline apps/src/javalab/JavalabView.jsx]: you don't need the constructor here. you can initialize state in the root of the component like this:

```javascript
state = {
  loading: true,
  loadSuccess: null
};
```

File: apps/src/javalab/JavalabEditor.jsx
Comment (maddiedierker) [inline apps/src/javalab/JavalabEditor.jsx:53]: if these 3 functions are arrow functions, you don't need to manually bind them -- they'll automatically be bound to `this`

Comment (maddiedierker) [reply apps/src/javalab/JavalabEditor.jsx:53]: if you do that, you can also get rid of the constructor and set the initial component state in the root of the component

File: apps/src/javalab/JavalabFileManagement.js
Comment (molly-moen) [reply apps/src/javalab/JavalabFileManagement.js]: 🤦‍♀️ good call

File: apps/src/javalab/JavalabEditor.jsx
Comment (molly-moen) [reply apps/src/javalab/JavalabEditor.jsx:53]: the reason I bound them here was because of this part of the [style guide](https://github.com/airbnb/javascript/tree/master/react#methods) Jessie mentioned--apparently using arrow functions for event handlers can slow things down.

Comment (maddiedierker) [reply apps/src/javalab/JavalabEditor.jsx:53]: TIL! thanks for sharing that -- you can ignore that comment 😄 you can still set the initial state outside of the constructor, but that seems less necessary if we'll be keeping the constructor anyways

Comment (molly-moen) [reply apps/src/javalab/JavalabEditor.jsx:53]: do we have a preference for initializing state inside/outside the constructor? Didn't know before today you could initialize it outside :)

Comment (maddiedierker) [reply apps/src/javalab/JavalabEditor.jsx:53]: i don't think we have an explicit preference. what you have aligns with what i've seen in our codebase -- set initial state in the component root if there isn't a constructor, otherwise set initial state inside the constructor

**Review by maddiedierker — APPROVED** [review]
great work molly, it's so exciting to see the progress in javalab! 🎉

**Pull request merged**
